### PR TITLE
core: (trait) Implement Lazy Class to workaround circular dependency

### PR DESCRIPTION
**Pull request by kingiler**
Address #1218

~~`LazyTrait` is implemented to postpone evaluation of provided traits.~~
`Lazy` class is implemented to postpone evaluation of provided objects.

### CODE REVIEW

File: xdsl/traits.py [hunk 1]
```diff
@@ -32,6 +33,23 @@ def verify(self, op: Operation) -> None:
...
+    """A trait that postpone evaluation of other traits. Can be used to workaround circular dependency."""
+
+    parameters: tuple[Callable[[], OpTrait | tuple[OpTrait, ...]], ...]
+
+    def __init__(self, *parameters: Callable[[], OpTrait | tuple[OpTrait, ...]]):
+        super().__init__(parameters)
+
+    def verify(self, op: Operation) -> None:
+        for func in self.parameters:
+            traits = func()
```
Comment (superlopuh) [inline xdsl/traits.py:46]: I think we might as well cache this to `self._traits`

**Review by superlopuh — APPROVED** [review]
Excellent, thank you

**Review by math-fehr — CHANGES_REQUESTED** [review]
The current solution would not work with `has_traits`.
Can we instead make the entirety of `traits` lazy? I feel that this would reduce the complexity overall.
We would write `traits = lazy(lambda: frozenset([...]))`, and any time we would access `traits`, it would compute the traits and cache the results.
I think it might be possible to do that with a `lazy` function that cheat at pyright level, but we could also probably do that with descriptors https://docs.python.org/3/howto/descriptor.html (I'm just not sure if they are available for class fields)

File: xdsl/traits.py (hunk 1 above)
Comment (kingiler) [reply xdsl/traits.py]: Cached in recent commit.

File: xdsl/ir/core.py [hunk 2]
```diff
@@ -583,7 +583,7 @@ class Operation(IRNode):
     _prev_op: Operation | None = field(default=None, repr=False)
     """Previous operation in block containing this operation."""
 
-    traits: ClassVar[frozenset[OpTrait]]
+    traits: ClassVar[LazyTrait]
```
Comment (AntonLydike) [inline xdsl/ir/core.py:586]: Not sure I'm on board with this?

**Comment by superlopuh** [comment]
Hmm, I kind of liked it more before. I'm not sure exactly what implementation @math-fehr had in mind, but it was nice that the thing stayed a frozenset, and that only lazy things had to be lazy

**Comment by kingiler** [comment]
I'm not sure if you mean the implementation in #1375? It would be minimal and less intrusive, yet more generic with dirty hacks.

**Comment by math-fehr** [comment]
@superlopuh What I have in mind is:
* We define traits with something like `traits = traits_def()`, or `traits = OpTraits(...)`
* We can either pass a `lambda` to `OpTraits`, or pass directly a `Sequence` of traits. In the end, it will always be a `frozenset` in the background, so that's fine
* I don't think a generic `Lazy` descriptor would be useful here, since we want to have a sequence of elements. So I would just have a custom class here for that.

**Comment by math-fehr** [comment]
Closing this in favor of #1529

**Pull request closed without merging**